### ISSUE TITLE
prohibit named objects on the > 1 level from abstract object

### ISSUE THREAD

**Issue by maxonfjvipon** [objectionary/lints#291]
This is EO code with syntax sugar:
```
[] > foo
  x > @
    y > a
      z > b
```

The canonical form of such code: 
```
[] > foo
  z > b
  y $.b > a
  x $.a > @
```

The same structure will be in XMIR. With sugar:
```xml
<objects>
  <o base="x" name="@">
    <o base="y" name="a">
      <o base="z" name="b"/>
    </o>
  </o>
</objects>
```

The canonical:
```xml
<objects>
  <o base="x" name="@">
    <o base="$.a"/>
  </o>
  <o base="y" name="a">
    <o base="$.b"/>
  </o>
  <o base="z" name="b"/>
</objects>
```

The key point - only the FIRST-level children of abstract object (or `<objects/>` element) may have `@name` attribute, any `@name` attributes found deeper must be prohibited

The severity level is `critical`

**Comment by yegor256** [comment]
@maxonfjvipon can you show bad code first (with a warning) and then its better alternative (without a warning)?

**Comment by maxonfjvipon** [comment]
@yegor256 this is incorrect (`@name` attribute is met on level 3 and 4 - `a`, `b`, `abc`)
```xml
<objects>
  <o base="x" name="@">
    <o base="y" name="a">
      <o base="z" name="b"/>
    </o>
  </o>
  <o name="foo">
    <o base="xyz">
      <o base="num" name="abc"/>
    </o>
  </o>
</objects>

```
This is correct, no `@name` attributes on level >= 1
```xml
<objects>
  <o base="x" name="@">
    <o base="$.a"/>
  </o>
  <o base="y" name="a">
    <o base="$.b"/>
  </o>
  <o base="z" name="b"/>
  <o name="foo">
    <o base="xyz">
      <o base="$.abc"/>
    </o>
    <o base="num" name="abc"/>
  </o>
</objects>
```

We start count levels from closes abstract object or `<objects/>` element

**Comment by yegor256** [comment]
@maxonfjvipon what is the motivation for this rule?

**Comment by maxonfjvipon** [comment]
@yegor256 this is just an incorrect structure of xmir, we can't work with it, it's not a "canonical" form
```
<objects>
  <o base="x" name="@">
    <o base="y" name="a">
      <o base="z" name="b"/>
    </o>
  </o>
  <o name="foo">
    <o base="xyz">
      <o base="num" name="abc"/>
    </o>
  </o>
</objects>
```

**Comment by yegor256** [comment]
@maxonfjvipon but this code is valid:

```
[] > foo
  x > @
    y > a
      z > b
```

This issue is only about XMIR, not EO?

**Comment by maxonfjvipon** [comment]
@yegor256 it's sugared EO, after full parsing pipeline it'll look different

**Comment by yegor256** [comment]
@maxonfjvipon do you suggest making this code illegal?

```
[] > foo
  x > @
    y > a
      z > b
```

**Comment by maxonfjvipon** [comment]
@yegor256 It's legal at the level of EO, but I think it's illegal at the level of XMIR, after full parsing with all of the transformations, it'll look like this:
```
<objects>
  <o base="x" name="@">
    <o base="$.a"/>
  </o>
  <o base="y" name="a">
    <o base="$.b"/>
  </o>
  <o base="z" name="b"/>
</objects>
```
I think we should not allow the next XMIR (for example if someone generate it not from EO)
```
<objects>
  <o base="x" name="@">
    <o base="y" name="a">
      <o base="z" name="b"/>
    </o>
  </o>
</objects>
```
This XMIR requires additional transformations (`vars-float-up`)

**Comment by yegor256** [comment]
@maxonfjvipon got it, make sense
